docs/source/scheduling-state.rst

```diff
@@ -70,153 +76,197 @@ Each of the following is a dictionary keyed by task name (described below):
+   workers.
+
+* **worker_restrictions:** ``{key: {worker addresses}}``:
+
+   A set of complete worker addresses per key of where that key can
+   be run.  Usually this is empty unless a key has been specifically
+   restricted to only run on certain workers.
+
+* **resource_restrictions:** ``{key: {resource: quantity}}``:
+
+   Resources required by a task, such as ``{'gpu': 1}`` or ``{'memory': 1e9}``.
+   These are user-defined names and are matched against the contents
+   of the ``worker_resources`` dictionary.
+
+* **loose_restrictions:** ``{key}``:
+
+   Set of keys for which we are allowed to violate restrictions (see above)
+   if no valid workers are present and the task would otherwise go into the
+   ``unrunnable`` set.  In other words, if a key is in ``loose_restrictions``,
+   then its restrictions become mere preferences, otherwise they are mandatory.
+
+* **who_wants:** ``{key: {client}}``:
+
+   When a client submits a graph to the scheduler it also specifies which
+   output keys it desires.  Those keys are tracked here where each desired key
+   knows which clients want it.  These keys will not be released from memory
+   and, when they complete, messages will be sent to all of these clients that
+   the task is ready.
+
+* **wants_what:** ``{client: {key}}``:
+
+   The transpose of ``who_wants``.
+
+Task state flow
+'''''''''''''''
+
+These state variables reflect the current status of a task and may
+get updated at each state transition.
 
 * **task_state:** ``{key: string}``:
 
-    The ``task_state`` dictionary holds the current state of every key.
-    Current valid states include released, waiting, no-worker, processing,
-    memory, and erred.  These states are explained further below.
+   The ``task_state`` dictionary holds the current state of every key.
+   Current valid states include ``released``, ``waiting``, ``no-worker``,
+   ``processing``, ``memory``, and ``erred``.  These states are explained
+   :ref:`further below <task-states>`.
 
-* **priority:** ``{key: tuple}``:
+* **waiting and waiting_data:** ``{key: {keys}}``:
 
-    The ``priority`` dictionary provides each key with a relative ranking.
-    This ranking is generally a tuple of two parts.  The first (and dominant)
-    part corresponds to when it was submitted.  Generally earlier tasks take
-    precedence.  The second part is determined by the client, and is a way to
-    prioritize tasks within a large graph that may be important, such as if
-    they are on the critical path, or good to run in order to release many
-    dependencies.  This is explained further in :doc:`Scheduling Policy
-    <scheduling-policies>`
+   These dictionaries are a subset of ``dependencies`` and ``dependents``
+   respectively, as they only track keys that are still in play.
 
-    A key's priority is only used to break ties, when many keys are being
-    considered for execution.  The priority does *not* determine running order,
```

Comment:
"The priority does *not* determine running order" sounds wrong to me, otherwise what would it be used for? The goal is to start computing some tasks earlier than others...